### DISCUSSION TITLE
request - Usage tracker

### DISCUSSION THREAD

**Discussion by TimyIsCool** [Kilvoctu/aiyabot#109] [Ideas]
Hi, i was wondering if some sort of GPU, CPU, and RAM usage along with temp and power draw tracker would be nice, like in the 'playing' part or while its generating it shows somewhere, i would find it useful
also a 'cached in vram' (idk if it has a cached in vram or if it only loads one model into vram) and 'cached in ram' when looking at all models would be nice
EDIT: Also a it/s counter would be cool and maybe showing the img say every 5 steps till its done

**Comment by TimyIsCool** [comment]
also unrelated (wanted to save opening up another issue) 
Im getting a `Exceeded maximum of 1 images! This is the best I can do...` how do i increase that to say 4?
EDIT: Nvm found how to do this

**Comment by Kilvoctu** [comment]
These would be interesting things to have, but there are some problems which I think would cause some issues with implementing this feature 🤔 I'll go through each thing and give my thoughts.

> GPU, CPU, and RAM usage along with temp and power draw tracker

AIYA is a separate app from Web UI, and I can't assume she'll be on the same machine as what's running Web UI. This means to me it's not practical to give her a native function to monitor machine stats, since she might be on separate machine. I also am not sure I want AIYA to be monitoring user's PC anyways.  
The other method would be pulling these stats from Web UI, through the API. The issue is the API only provides the memory usage (at `/sdapi/v1/memory`); not much particularly useful.  
So at this point, it may be more practical to have a dedicated PC stats bot that runs on same machine as Web UI. Something like https://github.com/3ddelano/pc-statbot, or maybe you can find a better one.

>  like in the 'playing' part or while its generating it shows somewhere

In the past I played around with updating the bot activity status thing. Well, Discord really doesn't like when bots are updating that every second or few seconds, lol.  
So I think the next best place is the initial "wait for me" message. I'll look into this sometime to see if it can be done and is practical. I'll add the idea to #85 later.  
A possible issue is that AIYA's queue is independent from Web UI's queue, so I'm not sure if I would be able to report on the correct task, and the stats provided by API endpoint `/sdapi/v1/progress` are a bit lacking.

> showing the img say every 5 steps till its done

I like this feature, actually (I use it in Web UI). And I know for example, Midjourney bot does it, but AIYA will not do so. Two reasons:
- The code around AIYA's queuing and posting to Discord is very convoluted, and I'd have to rewrite the whole thing to implement this.
- There is a rather significant performance impact. Sending images to Discord for some reason takes a long time. There's typically a few seconds after an image generation is done before it gets posted to Discord. I just recently refactored code to improve performance in this regard in #89.  

So for me, I feel like the value of seeing a preview doesn't warrant a rewrite of foundational code and performance loss.

**Comment by Kilvoctu** [comment]
> EDIT: Also a it/s counter would be cool and maybe showing the img say every 5 steps till its done

Heya, I saw this fork you may want to try out. They have an implementation of in-progress stats and preview pics    
https://github.com/phoenix-the-uncreative/aiyabot  
The feature in practice sorta reminds me why I'm not pursuing this functionality lol, but it seems to work all right enough and could be good for your use case. It'll look like this  
![image](https://user-images.githubusercontent.com/2993060/219263570-7de90bb2-cb9d-4e69-9c84-3c2c41e1e810.png)

**Reply by TimyIsCool** [reply]
Looks cool, will have a look once im home
